bin/bl-recent-files-pipemenu

```diff
@@ -81,7 +89,16 @@ end
 -- @return XML object if successful, otherwise nil
 local function add_items_from_xbel(m, path, reverse_output, cnt, override_cmd)
   local function clean_executable_name(s)
-    return s:match("([^'^%s]+)")
+    s = s:match("^'%s*(.-)%s*'$")
```

Comment:
When handling arbitrary command names, how about resolving any "command" to an absolute path by looking it up in path, returning only a valid path to a file system object, no cleaning required, and putting only absolute executable names in the pipemenu? Of course, proper quoting (for jgmenu, openbox) would still be required.

Especially, because of jgmenu's use of popen, the resulting command string in the pipemenu must not have a command ending with `\`, because that would join the arguments to the executable, in terms of shell.

Reply:
Yes quoting is (as always) tricky. I was unsure how best to handle the commands and agree it feels a bit fragile the way I had it. I thought splitting the *exec* attribute into words, and quoting each word separately, might be overkill. Maybe though, that's what should be done?

The specific use case which brought this up was *meld*, which writes bookmark entries like this:
```
  <bookmark href="file:///home/john/.local/share/meld/recent-9vquzsf9.meldcmp" added="2021-04-17T08:10:17Z" modified="2021-04-19T05:38:12Z" visited="2021-04-17T08:10:17Z">
    <title>bl-recent-files-pipemenu-tweaked-skipnonexistant : bl-recent-files-pipemenu</title>
    <desc>File comparison
file:///home/john/projects/lua/bl-recent-files-pipemenu-tweaked-skipnonexistant, file:///home/john/git/bunsen/bunsen-pipemenus/bin/bl-recent-files-pipemenu</desc>
    <info>
      <metadata owner="http://freedesktop.org">
        <mime:mime-type type="application/x-meld-comparison"/>
        <bookmark:applications>
          <bookmark:application name="Meld" exec="&apos;/usr/bin/meld --comparison-file %u&apos;" modified="2021-04-19T05:38:12Z" count="2"/>
        </bookmark:applications>
        <bookmark:private/>
      </metadata>
    </info>
  </bookmark>
```
Such entries, if executed correctly, allow the file comparison to be re-run, which is quite useful, but the `--comparison-file` argument has to be added to the command line. (A further enhancement would be to use the \<title\> tag as the menu label, instead of the filename.)

>the resulting command string in the pipemenu must not have a command ending with \, because that would join the arguments to the executable, in terms of shell.

OK, so if all words were single-quoted , maybe it would be safe? `'oddcmd\' '--arg'`I don't think executable names ending in a backslash can actually exist, but I guess we should cope with a malformed .xbel file? At least not execute anything suspicious.

Reply:
> how about resolving any "command" to an absolute path by looking it up in path, returning only a valid path to a file system object, no cleaning required, and putting only absolute executable names in the pipemenu?

This commit: https://github.com/BunsenLabs/bunsen-pipemenus/pull/62/commits/54bcf54c01375df2621f1dbfa43c4d8fd0ff6f5b is an attempt to implement that. 

I couldn't find any Lua library function like the unix *which*, so made this function:
```
  local function which(cmd)
    local path = os.getenv("PATH")
    if cmd:match("^/") and posix.access(cmd, x) then -- absolute path
      local cdir = posix.dirname(cmd)
      for p in path:gmatch("[^:]+") do
        if p == cdir then return cmd end
      end
    elseif not cmd:match("/") then -- reject relative path fragments
      for p in path:gmatch("[^:]+") do
        local fp = p.."/"..cmd
        if posix.access(fp, x) then return fp end
      end
    end
    return nil
  end
```
While looking up single-word executables in $PATH, anything absolute is permitted only if it lies in one of the directories in $PATH. I'm not 100% sure whether ruling out custom filepaths is correct, but since this is a convenience utility I thought just skipping dubious menu items was a reasonable policy.

@2ion I'd appreciate your input here, and of course approval if this branch were to be merged in.